ynput/ayon-usd
# LakeFS support for binary distribution

# TODO 

- [x] Setting to acknowledge that this feature is not production-ready 
- [x] don't start the addon if the setting is not switched on.  

- [x] Windows title for download USD window
- [x] Settings titles without underscores but with spaces
- [x] Shorter addon name
- [x] version info the addon class
- [x] update settings so that lake fs can use the new naming system
- [x] update the find resolver system so that it will find the resolver in the new layout
    - [x] win
- [x] Figure out why the prelaunch hook does not download under windows 
- [x] Currently, the download system does not take into account if the JSON has an entry but the file/folder is missing
- [x] rename the `usd-24.03_win64_py39.zip` to `usd-24.03_windows_py39.zip` to allow ayon_usd to find it on Windows. (we will unify the naming via another PR)
- [x] convert the string evaluation off settings keys into a tuple evaluation
- [x] update ayon_bin_bridge_client to the newest version 
- [x] test if the new version of bin_client will run under Windows (the ne ci tests do run)
- [x] better error message when get_element_info() cant find the element (check if element return is empty then print that no element was found)
- [x] better error message when no resolver could be found (include: app_name)
- [x] don't cache server settings (`PreLaunchHook` instances have access to the `self.data["project_settings"]`)
- [x] better create package py 
    - [x] git submodule init (wont fix for alpha, setup should change to AyonAutomator)
    - [x] fix client site downloads folder zipping 
- [x] create package should ignore downloads, not delete it  


## Testing Notes 
### Getting started:

- Create ayon-usd package 
- Upload to AYON server (this addon adds new settings to the server)
- Request credentials (currently only on Ynput-Discord) for the LakeFS distribution.
- Enter the credentials into the addon settings
    - ayon+settings://ayon_usd/LakeFs_Settings/access_key_id
    - ayon+settings://ayon_usd/LakeFs_Settings/secret_access_key
- Switch the branch from main to pinning-suport (this is because there is a compile error in Main that is fixed in this branch https://github.com/ynput/ayon-usd-resolver/issues/21)
    - this setting ayon+settings://ayon_usd/LakeFs_Settings/ayon_usd_lake_fs_server_repo
    - to this value: lakefs://ayon-usd/pinning-suport/ (default will be lakefs://ayon-usd/main/)
- Launch AYON
- Launch AYON USD Resolver supported DCC (Unreal, Houdini, Maya, _AyonStandalone_ is **not** yet implemented).

Optionally, you can set the log level (ayon+settings://ayon_usd/Ayon_UsdResolver_Settings/ayon_log_lvl) to Info. When the resolver is initialised, this will print some init output in the Houdini console (Windows) or terminal (Linux).  

### What to test:

- AyonUsdLib should be downloaded on Launcher startup (all downloads are stored in the addon dir. You can find your addon dir by clicking the version number in the tray menu (this will open Ayon info). Then, you can open Environments and take the path from the AYON_ADDONS_DIR. Use this to navigate to the install dir of your ayon-usd-addon, and you will find a downloads dir)
- Resolver should get installed on Software Startup (the resolver will be downloaded in the same place as the UsdLib)
- Resolver should work in Houdini 19.5.xxx and 20.0.xxx, Maya2024, Maya2025, Unreal5.4

In case you don't have a loader that uses Ayon URI, do the following: 
- Open Ayon Web view. 
- Go to Projects-Browser, select a published USD asset 
- select a representation and copy the entity URI
    - the URI looks like this ayon+entity://{ProjectName}/AssetName?product=usdAsset&version={VersionNumber}&representation=usd 
    - the representation should be USD
- Launcher Software Via Ayon Launcher 
  - Houdini 
    - use a reference or sublayer node and put the URI into the file path. 
  - Maya
    - Click Create/Universal Scene Description (USD)/Stage With New Layer
    - Your outliner's new stage, "Object", has a Stage-Source/File (Attribute Editor). Place the URI into the File Filed. 
  - Unreal
    - create a Usd-Stage-Actor (you will need to activate the usdImport plugin in Unreal)
    - create a USD file where you reference a ayon uri (you can create this in Houdini or in Maya) load the USD file via its file path (see known issues if you want to know why)

## Known Issues 
- The downloaded file is currently blind to updates on lakeFs. This means that when you change your branch or upload a new resolver, this system will not redownload the resolver. (we should talk about how we want to fix this, (there was the idea to have build tags on the LakeFs zips and compare them, or we could write a JSON file with the data when we downloaded the resolver to see if a resolver with a newer time stamp exists. @antirotor @BigRoy @dee-ynput not sure any more who I should ask for a Greenligth on that decision))
- in unreal, when you create a stage actor, the resolver will not be passed the URI, but instead, it will be passed the project path + the Uri `/home/lyonh/Documents/Unreal Projects/Dell/ayon+entity:/Usd_Base/trees?product=usdtree_106&version=v001&representation=usd` this causes the resolve to fail. (there is no decision how we should fix this)


### This Pr handles implementation of LakeFs into the Ayon-Usd addon.

# Components 

## Ui
The ui is mainly build on the Server side, the addon itself dose not have any ui to configure 
General Ui
![image](https://github.com/ynput/ayon-usd/assets/157467381/419e44ab-005f-4861-b888-2ac17b1b63fa)
LakeFs config ui (i would like to reduce this to a point where the lakeFs Package has a file that informs about the internal structure, maybe via an package_layout.json or similar so that we only need an alias config where we could say (key for Houdini is Hou20.0 but app_name is houdini20.0) but this will probably need to wait for this pr [ayon-usd_lakefs-package_build](https://github.com/ynput/ayon-usd_lakefs-package_build/issues) )
![image](https://github.com/ynput/ayon-usd/assets/157467381/7272f3c1-2951-43c5-93f1-5ba1c697e4a9)

## Addon
the addon pulls the UsdLib by requesting the layout from the lakeFs server and then comparing the available UsdLibs with the platform name. (all of this should / could be handled by the package_layout.json at some point )

the resolver are pulled the same way but using the app_name and the mapping defined in the server settings. 
Application overwrites will allays take control in case they are defined for a specific app_name in all other cases the LakeFs paths will be used in combination with the current defined branch.
